### ISSUE TITLE
Loading knitr when using checkpoint + rmarkdown fails when snapshot is too old

### ISSUE THREAD

**Issue by fzenoni** [RevolutionAnalytics/checkpoint#264]
I am experiencing a weird problem when loading the `knitr` package inside an RMarkdown document.
First of all, here is my `sessionInfo()`

```r
R version 3.4.3 (2017-11-30)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 16.04.3 LTS

Matrix products: default
BLAS: /usr/lib/openblas-base/libblas.so.3
LAPACK: /usr/lib/libopenblasp-r0.2.18.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
[1] compiler_3.4.3 tools_3.4.3    yaml_2.1.16  
```

If the snapshot is recent (e.g. 2018-01-10) then there is no problem. If it is older (e.g. 2017-09-01), then the Knit rendering fails by throwing the following error.

```
Error in value[[3L]](cond) : 
  Package 'knitr' version 1.18 cannot be unloaded:
 Error in unloadNamespace(package) : namespace 'knitr' is imported by 'rmarkdown' so cannot be unloaded
Calls: <Anonymous> ... tryCatch -> tryCatchList -> tryCatchOne -> <Anonymous>
```

Here are minimal reproducible scripts for

* the old snapshot ->https://github.com/fzenoni/old_snapshot/blob/master/analysis/main.Rmd
* and the new snapshot -> https://github.com/fzenoni/new_snapshot/blob/master/analysis/main.Rmd

the only difference between them being the `package_date` variable on Line 19.

The apparent complex appearance of the packages linked above derives from [this template](https://github.com/grssnbchr/rddj-template) for reproducible research, which in turn derives from this [checkpoint vignette](https://github.com/RevolutionAnalytics/checkpoint/blob/master/vignettes/archive/using-checkpoint-with-knitr.Rmd).

**Comment by grssnbchr** [comment]
As the author of the mentioned template, I tried to reproduce your error with both 3.4.2 and 3.4.3, and was unsuccesful. So it worked for both R versions and both snapshot dates. 

I have also seen errors like the one you report before – often the solution lies in *always* restarting R before changing the snapshot date and running it again (in RStudio: Session > Restart R and Run All Chunks).

Sometimes, even restarting RStudio altogether or killing the rsession processes with `killall rsession` in the shell is necessary, but these cases are seldom.

A more radical measure could be to delete the .checkpoint folder altogether and start from scratch.

**Comment by fzenoni** [comment]
Thank you @grssnbchr for your input. Could you specify on what OS have you tried the code? I noticed that Windows has a much higher "success rate".

**Comment by fzenoni** [comment]
As documented in the [README](https://github.com/RevolutionAnalytics/checkpoint/blob/master/README.md), to replace `knitr` with `rmarkdown` is a valid workaround, when one for instance wants to execute the following chunk:

```r
dt <- mtcars[1:5, 1:6]
knitr::kable(dt, "html")
```

This is successfully rendered regardless of the snapshot, if I do Session > Restart R and Run All Chunks.
Why does it fail then when I press the Knit button?

```
Output created: main.html
Error in sprintf("^%s[0-9a-f]+[.]html$", tmpfile_pattern) : 
  object 'tmpfile_pattern' not found
Calls: <Anonymous> -> clean_tmpfiles -> unlink -> list.files -> sprintf
Execution halted
```

**Comment by grssnbchr** [comment]
I tried it in Ubuntu 16.04. 

I also suggest restarting R between switching from interpreter mode (Run All Chunks) to knitting a document.

Also, why do you even need to load `knitr` or `rmarkdown` explicitly? If you look at my template (which can be knitted without problems), **neither** of them is loaded: https://github.com/grssnbchr/rddj-template/blob/master/analysis/main.Rmd#define-packages. `knitr::kable` works nonetheless.

**Comment by fzenoni** [comment]
> Also, why do you even need to load knitr or rmarkdown explicitly? If you look at my template (which can be knitted without problems), neither of them is loaded: https://github.com/grssnbchr/rddj-template/blob/master/analysis/main.Rmd#define-packages. knitr::kable works nonetheless.

I still don't know how my previous scripts worked for you, but at this moment this is the best answer for me.
Thank you very much for taking the time to do the tests.

**Comment by grssnbchr** [comment]
I think I know now why you had this problem. I encountered a similar one. It is actually related to the fact that, if you knit your document (e.g. in RStudio), the *globally* installed `rmarkdown` / `knitr` packages with all their dependencies are used (just as a global installation of `checkpoint` is used). 

They might not work well together with stuff that you install *locally* within your checkpoint process. So for example, if you load `knitr` *locally*, a global `knitr` with a *different version* is already loaded. However, you didn't have these problems with a then recent package date. That's because, presumably, your *local* installation (with package date 2018-01-10) was identical to your *global* installation of knitr. 

All of this is a clear drawback of my approach, or in other words: while checkpoint is perfectly fine for reproducing "ordinary" R scripts, it has problems with rendering as RMarkdown. There should be a way to render the document with a *local* installation of `rmarkdown`, coming from the same package date as the other packages, however, I haven't found a way to do this.

**Comment by fzenoni** [comment]
It makes sense. As long as `knitr` and `rmarkdown` functions stay consistent, not loading the package in the `checkpoint` environment is a good enough workaround. Thank you for the follow up.

**Comment by grssnbchr** [comment]
I think I found a way to knit the RMarkdown using a *local* `rmarkdown` package from the same CRAN package date as the packages used in the RMarkdown script. However, knitting *inside* RStudio is no longer possible / recommended, I use a shell script for knitting now. This shell script sets the  `R_LIBS_USER` to the current `.checkpoint` directory before knitting the script, this way, the *local* installation of `rmarkdown` is used. I've updated https://github.com/grssnbchr/rddj-template. The shell script is here: https://github.com/grssnbchr/rddj-template/blob/master/knit.sh, lines 41 and 42 are most interesting.
